### ISSUE TITLE
Question: Can we convert number to string in expression? or have C style strings in string operations?

### ISSUE THREAD

**Issue by reportaman** [andmarti1424/sc-im#501]
Is there a way to convert a number to its string representation, ideally one would be able to specify full C-style string formatting option.

Something like: ``@ntos("format string",number)``

It could perhaps just be a generic cstring option

``@str("format string",comma separated list of arguments)``

**Comment by andmarti1424** [comment]
No. I believe its not possible. Asumming someone you create that function, it would return a string.
Do you think it would be useful to save a number in a string representation?

**Comment by reportaman** [comment]
Yes very useful! The case I was thinking of: For functions that take ranges like ``=@avg(A1:A37)``, one could instead get the substrings A, 1, 37 from other functions like ``@coltoa()`` and proposed ``@ntos``. I think it could even work with a simple, and similar implementation as ``@coltoa`` to instead get row number as string to use in concatenation operation & feed it to other functions like ``@avg``. The advantage of having this would be: one can write expressions which specify relative position instead of absolute position. Then even if someone cut-copy-pastes parts of existing spreadsheet, one need not update the expressions!

Also, a breaking change could be to deprecate ``@coltoa``, and instead have ``@cols`` that does that ``@coltoa`` does and ``@rows`` would be the only new addition (and returns string "12" for numeric ``@myrow`` value of ``12.00`` )

**Comment by reportaman** [comment]
Copy from our discussion on https://github.com/andmarti1424/sc-im/issues/497#issuecomment-786607750:

The other implementation of SC seems to have a ``@fmt`` function that does what I was asking for in ``@ntos``. Not sure how easy it is for you to just pull in that function definition, and apply it here. The only other feature it adds is mouse support, and I don't think there will be further changes there: https://github.com/n-t-roff/sc/issues/18

<img width="999" alt="Screen Shot 2021-02-26 at 4 02 53 AM" src="https://user-images.githubusercontent.com/71736629/109299032-4c880500-77e9-11eb-8b25-ff3abfd000a7.png">

**Comment by andmarti1424** [comment]
@reportaman Actually, @fmt works.
Try this:
@fmt("%6.3f",A0)

**Comment by reportaman** [comment]
@andmarti1424 Ok thanks! Please add it to ``:help`` document. Yes it seems to work, and here's the meaning of formats: https://www.gnu.org/software/libc/manual/html_node/Floating_002dPoint-Conversions.html

**Comment by reportaman** [comment]
Why does it not work in an expression like this: ``=@avg(@coltoa(@mycol)0:@coltoa(@mycol)@fmt("%.f",@myrow-1))``?

**Comment by andmarti1424** [comment]
I am really not sure.
@coltoa(@mycol)0
doest work either.
And this was what prevent me to finish my work on https://github.com/andmarti1424/sc-im/issues/451

**Comment by andmarti1424** [comment]
btw, this sure happens in old sc too. its inherited from there.

**Comment by reportaman** [comment]
@andmarti1424 Regarding ``@coltoa(@mycol)0``, this works: ``\"@coltoa(@mycol) # @fmt("%.f",@myrow)``

**Comment by reportaman** [comment]
@andmarti1424 And for range, this works too!

``\"@coltoa(@mycol) # "0:" # @coltoa(@mycol) # @fmt("%.f",@myrow)``

From cell C14, it prints: ``C0:C14``

**Comment by andmarti1424** [comment]
\"@coltoa(@mycol) # "0" works..
The thing is "0" is a string, but 0 of course no. And # concatenates 2 strings..
@fmt outputs a string as well..

**Comment by andmarti1424** [comment]
And @avg expects a range (defined in gram.y) not a string..

**Comment by reportaman** [comment]
@andmarti1424 If it is possible to accept either style, then something like this could work? Not sure if there are any caveats, and how difficult that is...

``=@avg(@coltoa(@mycol) # "0" # ":" # @coltoa(@mycol) # @fmt("%.f",@myrow))``

**Comment by andmarti1424** [comment]
Yes. If we manage to get a string like "B2" be interpreted as a (gram.y) var.
Then yes, a string with B2:B4 could be interpreted as a (gram.y) range without problem.
Its the same thing. And them both are (gram.y) var_or_range.
I just dont know how to achieve that in gram.y
@roman65536 may be?

**Comment by andmarti1424** [comment]
@reportaman take a look at the recent added `@evaluate` and `@sevaluate` functions

**Comment by andmarti1424** [comment]
@reportaman and after todays commits, also `@getent`

**Comment by reportaman** [comment]
@andmarti1424 Btw you might consider standardizing the interface. For range functions now``@getent`` works to operate on a range relative to the position of a cell. But for non-range functions like ``@nval()``, one needs to use a different interface to mean the same thing. For instance ``@nval(@coltoa(@mycol-4),@myrow)`` to get value relative to current cell. Ideally ``@getent`` could be used everywhere user wants to code logic relative to existing cell. So this would ideally look like ``@nval(@getent(@mycol-4,@myrow))``.

**Comment by andmarti1424** [comment]
Yes. Using string expression as @getent first parameter was first idea. But I thought would be better to get it working first and then change it.
@nval takes string expression as first parameter. So, @nval(@getent(..)) wont work, in the same way @nval(B0) wont work either.
